phoenix-core/src/main/java/org/apache/phoenix/iterate/OffsetResultIterator.java

```diff
@@ -32,17 +36,27 @@
 public class OffsetResultIterator extends DelegateResultIterator {
     private int rowCount;
     private int offset;
+    private long pageSizeMs = Long.MAX_VALUE;
 
     public OffsetResultIterator(ResultIterator delegate, Integer offset) {
         super(delegate);
         this.offset = offset == null ? -1 : offset;
     }
 
+    public OffsetResultIterator(ResultIterator delegate, Integer offset, long pageSizeMs) {
+        this(delegate, offset);
+        this.pageSizeMs = pageSizeMs;
+    }
     @Override
     public Tuple next() throws SQLException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
         while (rowCount < offset) {
-            if (super.next() == null) { return null; }
+            Tuple tuple = super.next();
+            if (tuple == null) { return null; }
             rowCount++;
+            if (EnvironmentEdgeManager.currentTimeMillis() - startTime >= pageSizeMs) {
```

Comment:
I'm unclear why we're enforcing the time limits both in the FIlter/coprocs and also at the client level in the Iterators/Tuples. Do we need the latter in queries processed primarily on the client side? I'd expect the Iterator to be checking the Tuple returned to the client to see if it was the magic Dummy result, rather than checking the time itself.

Reply:
OffsetResultIterator is also used on the server side. When it is used on the client side, pageSizeMs is set to Long.MAX_VALUE in the constructor.